Source/WebCore/html/HTMLMaybeFormAssociatedCustomElement.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "HTMLMaybeFormAssociatedCustomElement.h"
+
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(HTMLMaybeFormAssociatedCustomElement);
+
+using namespace HTMLNames;
+
+HTMLMaybeFormAssociatedCustomElement::HTMLMaybeFormAssociatedCustomElement(const QualifiedName& tagName, Document& document)
+    : LabelableElement(tagName, document)
+{
+    ASSERT(Document::validateCustomElementName(tagName.localName()) == CustomElementNameValidationStatus::Valid);
+}
+
+HTMLMaybeFormAssociatedCustomElement::~HTMLMaybeFormAssociatedCustomElement() = default;
+
+Ref<HTMLMaybeFormAssociatedCustomElement> HTMLMaybeFormAssociatedCustomElement::create(const QualifiedName& tagName, Document& document)
+{
+    return adoptRef(*new HTMLMaybeFormAssociatedCustomElement(tagName, document));
+}
+
+bool HTMLMaybeFormAssociatedCustomElement::isFormListedElement() const
+{
+    return canAccessFormAssociatedInternals();
+}
+
+bool HTMLMaybeFormAssociatedCustomElement::isValidatedFormListedElement() const
+{
+    return canAccessFormAssociatedInternals();
+}
+
+FormAssociatedElement* HTMLMaybeFormAssociatedCustomElement::asFormAssociatedElement()
+{
+    return canAccessFormAssociatedInternals() ? impl() : nullptr;
```

Comment:
impl() looks cryptic. Maybe asFormAssociatedCustomElement?

Reply:
Replaced with `formAssociatedCustomElementUnsafe()` in [3520b5d](https://github.com/WebKit/WebKit/pull/4988/commits/3520b5dbbb73fe3a3ee18339c67f0c38dc3b3c1c). `as*Element()` methods a bit like safe public APIs with additional checks.